arch/risc-v/src/common/riscv_misaligned.c

```diff
@@ -0,0 +1,448 @@
+{
+  uint8_t *dest;
+  uint8_t *src;
+  int      len;
+} rv_insn_ctx_t;
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: decode_insn_compressed
+ *
+ * Description:
+ *   Try to decode a compressed instruction to normal encoding
+ *
+ ****************************************************************************/
+
+static bool decode_insn_compressed(uintptr_t *regs, rv_insn_ctx_t *ctx)
+{
+#ifdef CONFIG_ARCH_RV_ISA_C
+  uint16_t in;
+  uint32_t imm = 0;
+  rv_insn_c_t insn;
+
+  /* Fetch instruction */
+
+  insn.insn = *(uint16_t *)regs[REG_EPC];
+
+  in = insn.insn & INSN_C_MASK;
+
+  switch (in & INSN_C_MASK)
+    {
+      case INSN_C_LW:
+        imm = insn.lw.imm2 << 2 | insn.lw.imm53 << 3 | insn.lw.imm6 << 6;
+        ctx->dest = (uint8_t *)&regs[REG_X8 + insn.lw.rd];
+        ctx->src  = (uint8_t *)regs[REG_X8 + insn.lw.rs1] + imm;
+        ctx->len  = 4;
+        break;
+
+      case INSN_C_LWSP:
+        imm = insn.lwsp.imm42 << 2 | insn.lwsp.imm5 << 5 |
+              insn.lwsp.imm76 << 6;
+        ctx->dest = (uint8_t *)&regs[insn.lwsp.rd];
+        ctx->src  = (uint8_t *)regs[REG_SP] + imm;
+        ctx->len  = 4;
+        break;
+
+      case INSN_C_LD:
+        imm = insn.ld.imm53 << 3 | insn.ld.imm76 << 6;
+        ctx->dest = (uint8_t *)&regs[REG_X8 + insn.ld.rd];
+        ctx->src  = (uint8_t *)regs[REG_X8 + insn.ld.rs1] + imm;
+        ctx->len  = 8;
+        break;
+
+      case INSN_C_LDSP:
+        imm = insn.ldsp.imm43 << 3 | insn.ldsp.imm5 << 5 |
+              insn.ldsp.imm86 << 6;
+        ctx->dest = (uint8_t *)&regs[insn.ld.rd];
+        ctx->src  = (uint8_t *)regs[REG_SP] + imm;
+        ctx->len  = 8;
+        break;
+
+      case INSN_C_SW:
+        imm = insn.sw.imm2 << 2 | insn.sw.imm53 << 3 | insn.sw.imm6 << 6;
+        ctx->dest = (uint8_t *)regs[REG_X8 + insn.sw.rs1] + imm;
+        ctx->src  = (uint8_t *)&regs[REG_X8 + insn.sd.rs2];
+        ctx->len  = 4;
+        break;
+
+      case INSN_C_SWSP:
+        imm = insn.swsp.imm52 << 2 | insn.swsp.imm76 << 6;
+        ctx->dest = (uint8_t *)regs[REG_SP] + imm;
+        ctx->src  = (uint8_t *)&regs[insn.swsp.rs2];
+        ctx->len  = 4;
+        break;
+
+      case INSN_C_SD:
+        imm = insn.sd.imm53 << 3 | insn.sd.imm76 << 6;
+        ctx->dest = (uint8_t *)regs[REG_X8 + insn.sd.rs1] + imm;
+        ctx->src  = (uint8_t *)&regs[REG_X8 + insn.sd.rs2];
+        ctx->len  = 8;
+        break;
+
+      case INSN_C_SDSP:
+        imm = insn.sdsp.imm53 << 3 | insn.sdsp.imm86 << 6;
+        ctx->dest = (uint8_t *)regs[REG_SP] + imm;
+        ctx->src  = (uint8_t *)&regs[insn.sdsp.rs2];
+        ctx->len  = 8;
+        break;
+
+      default:
+        _alert("Compressed: %x\n", insn.insn);
+        return false;
+    }
+
+  regs[REG_EPC] += 2;
```

Comment:
This looks a bit dubious, does this work ? I does this not result in an "Instruction address misaligned" fault ?

Reply:
Here is compression instruction, it look good to increase 2 bytes not 4 bytes.

Reply:
Yes, it's hard to recognize compressed instruction by a simple way, so it try to decode it first.